File: sunflower-core/src/main/java/org/tdf/sunflower/controller/EntryController.java
package org.tdf.sunflower.controller;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import com.sun.management.OperatingSystemMXBean;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.context.ApplicationContext;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.*;
import org.tdf.common.trie.Trie;
import org.tdf.common.types.Uint256;
import org.tdf.common.util.HexBytes;
import org.tdf.common.util.IntSerializer;
import org.tdf.common.util.RLPUtil;
import org.tdf.sunflower.AppConfig;
import org.tdf.sunflower.GlobalConfig;
import org.tdf.sunflower.consensus.poa.PoA;
import org.tdf.sunflower.facade.*;
import org.tdf.sunflower.net.Peer;
import org.tdf.sunflower.net.PeerServer;
import org.tdf.sunflower.state.Account;
import org.tdf.sunflower.state.AccountTrie;
import org.tdf.sunflower.state.Address;
import org.tdf.sunflower.sync.SyncManager;
import org.tdf.sunflower.types.*;

import java.lang.management.ManagementFactory;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.function.Function;
import java.util.stream.Collectors;

@RestController
@AllArgsConstructor
@RequestMapping("/rpc")
@Slf4j(topic = "rpc")
public class EntryController {
    private final AccountTrie accountTrie;

    @Qualifier("contractStorageTrie")
    private final Trie<HexBytes, HexBytes> contractStorageTrie;

    private final GlobalConfig config;


    private final PeerServer peerServer;

    private final TransactionPool pool;

    private final ObjectMapper objectMapper;

    private final RepositoryService repo;

    private final SyncManager syncManager;

    private final ConsensusEngine consensusEngine;

    private final Miner miner;

    private final ApplicationContext ctx;

    @SneakyThrows
    private Stat createState() {
        AppConfig c = AppConfig.get();

        Stat.StatBuilder builder = Stat.builder();
        OperatingSystemMXBean osMxBean = (OperatingSystemMXBean) ManagementFactory.getOperatingSystemMXBean();
        Block best;
        List<Block> blocks;

        try (RepositoryReader rd = repo.getReader()) {
            best = rd.getBestBlock();
            blocks = rd.getBlocksBetween
                (Math.max(0, best.getHeight() - 10), best.getHeight(), Integer.MAX_VALUE, false);
        }

        Uint256 totalGasPrice = Uint256.ZERO;
        long totalTransactions = 0;
        for (Block b : blocks) {
            for (int i = 0; i < b.getBody().size(); i++) {
                if (i == 0)
                    continue;
                totalGasPrice = totalGasPrice.plus(b.getBody().get(i).getGasPriceAsU256());
                totalTransactions++;
            }
        }

        String diff = "";

        return builder
            .cpu(osMxBean.getSystemLoadAverage())
            .memoryUsed(osMxBean.getTotalPhysicalMemorySize() - osMxBean.getFreePhysicalMemorySize())
            .totalMemory(osMxBean.getTotalPhysicalMemorySize())
            .averageGasPrice(
                totalTransactions == 0 ? Uint256.ZERO : totalGasPrice.div(Uint256.of(totalTransactions)))
            .mining(blocks.stream().anyMatch(
                x -> x.getBody().size() > 0 && x.getBody().get(0).getReceiveHex().equals(miner.getMinerAddress())))
            .currentDifficulty(diff).transactionPoolSize(0)
            .consensus(consensusEngine.getName())
            .gasPrice(c.getVmGasPrice().longValue())
            .build()
            ;
    }

    private <T> T getBlockOrHeader(RepositoryReader rd, String hashOrHeight, Function<Long, T> func,
                                   Function<HexBytes, T> func1) {
        Long height = null;
        try {
            height = Long.parseLong(hashOrHeight);
            Header h = rd.getBestHeader();
            while (height < 0) {
                height += h.getHeight() + 1;
            }
        } catch (Exception ignored) {

        }
        if (height != null) {
            final long finalHeight = height;
            return func.apply(height);
        }
        return func1.apply(HexBytes.fromHex(hashOrHeight));
    }

    @GetMapping(value = "/block/{hashOrHeight}", produces = MediaType.APPLICATION_JSON_VALUE)
    public BlockV1 getBlock(@PathVariable String hashOrHeight) {
        try (RepositoryReader rd = repo.getReader()) {
            return BlockV1.fromV2(getBlockOrHeader(rd, hashOrHeight, rd::getCanonicalBlock, rd::getBlockByHash));
        }
    }

    @GetMapping(value = "/header/{hashOrHeight}", produces = MediaType.APPLICATION_JSON_VALUE)
    public HeaderV1 getHeaders(@PathVariable String hashOrHeight) {
        try (RepositoryReader rd = repo.getReader()) {
            return HeaderV1.fromV2(getBlockOrHeader(rd, hashOrHeight, rd::getCanonicalHeader, rd::getHeaderByHash));
        }
    }

    @GetMapping(value = "/transaction/{hash}", produces = MediaType.APPLICATION_JSON_VALUE)
    public TransactionV1 getTransaction(@PathVariable String hash) {
        try (RepositoryReader rd = repo.getReader()) {
            return TransactionV1.fromV2(
                rd.getTransaction(HexBytes.fromHex(hash))
            );
        }
    }

    @GetMapping(value = "/account/{addressOrPublicKey}", produces = MediaType.APPLICATION_JSON_VALUE)
    public AccountView getAccount(@PathVariable String addressOrPublicKey) throws Exception {
        HexBytes addressHex = Address.of(addressOrPublicKey);
        try (RepositoryReader rd = repo.getReader()) {
            Account a = accountTrie.get(rd.getBestHeader().getStateRoot(), addressHex);
            if (a == null)
                a = Account.emptyAccount(Uint256.ZERO);
            return AccountView.fromAccount(addressHex, a);
        }
    }

    @GetMapping(value = "/peers", produces = MediaType.APPLICATION_JSON_VALUE)
    public PeersInfo peers() {
        return new PeersInfo(peerServer.getPeers(), peerServer.getBootstraps());
    }

    @GetMapping(value = "/miners", produces = MediaType.APPLICATION_JSON_VALUE)
    public List<HexBytes> miners() {
        return Collections.emptyList();
    }

    @GetMapping(value = "/orphan", produces = MediaType.APPLICATION_JSON_VALUE)
    public List<Block> orphans() {
        return syncManager.getOrphans();
    }

    @GetMapping(value = "/pool", produces = MediaType.APPLICATION_JSON_VALUE)
    public PagedView<TransactionV1> getPool(@ModelAttribute PoolQuery poolQuery) {
        return PagedView.empty();
    }


    @GetMapping(value = "/contract/{address}/abi", produces = MediaType.APPLICATION_JSON_VALUE)
    public Object getABI(@PathVariable("address") final String address) {
        return null;
    }

    @PostMapping(value = "/operations", produces = MediaType.APPLICATION_JSON_VALUE)
    public Object operations(@RequestBody JsonNode node) {
        String m = node.get("method").asText();
        switch (m) {
            case "export": {
//                List<Transaction> li = new ArrayList<>();
//                String publicKey = node.get("publicKey").asText();
//                HexBytes pk = HexBytes.fromHex(publicKey);
//                HexBytes address = Address.fromPublicKey(pk);
//                sunflowerRepository.traverseTransactions((h, t) -> {
//                    if (t.getFrom().equals(pk) || t.getTo().equals(address)) {
//                        li.add(t);
//                    }
//                    return true;
//                });
                return new ArrayList<>();
            }
        }
        throw new RuntimeException("invalid payload " + node.toString());
    }

    @GetMapping(value = "/stat", produces = MediaType.APPLICATION_JSON_VALUE)
    public Stat stat() {
        return createState();
    }

    // if thread node not receive farm base transaciont and gateway node restarted
    // needs to construct the transaction
    @GetMapping(value = "/farmBaseTransactions", produces = MediaType.APPLICATION_JSON_VALUE)
    public List<HexBytes> farmBaseTransactions() {
        PoA poa = (PoA) consensusEngine;
        return poa
                .farmBaseTransactions
                .asMap()
                .values()
                .stream()
                .map(RLPUtil::encode)
                .collect(Collectors.toList());
    }


    @AllArgsConstructor
    @Getter
    static class PeersInfo {
        List<Peer> peers;
        List<Peer> bootstraps;
    }

    @AllArgsConstructor
    @Getter
    static class AccountView {
        private final HexBytes address;

        // for normal account this field is continuous integer
        // for contract account this field is nonce of deploy transaction
        @JsonSerialize(using = IntSerializer.class)
        private final long nonce;

        // the balance of account
        // for contract account, this field is zero
        @JsonSerialize(using = IntSerializer.class)
        private final Uint256 balance;

        // for normal address this field is null
        // for contract address this field is creator of this contract
        private final HexBytes createdBy;

        // hash code of contract code
        // if the account contains none contract, contract hash will be null
        private final HexBytes contractHash;

        // root hash of contract db
        // if the account is not contract account, this field will be null
        private final HexBytes storageRoot;

        static AccountView fromAccount(HexBytes address, Account account) {
            return new AccountView(address, account.getNonce(), account.getBalance(),
                Address.empty(), account.getContractHash(),
                account.getStorageRoot());
        }
    }
}

File: sunflower-core/src/main/java/org/tdf/sunflower/controller/EntryController.kt
package org.tdf.sunflower.controller

import com.fasterxml.jackson.databind.annotation.JsonSerialize
import org.apache.commons.io.IOUtils
import org.slf4j.LoggerFactory
import org.springframework.http.MediaType
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import org.tdf.common.types.Uint256
import org.tdf.common.util.HexBytes
import org.tdf.common.util.IntSerializer
import org.tdf.common.util.hex
import org.tdf.sunflower.AppConfig
import org.tdf.sunflower.facade.RepositoryReader
import org.tdf.sunflower.facade.RepositoryService
import org.tdf.sunflower.net.Peer
import org.tdf.sunflower.net.PeerServer
import org.tdf.sunflower.state.Account
import org.tdf.sunflower.state.AccountTrie
import org.tdf.sunflower.state.AddrUtil
import org.tdf.sunflower.state.AddrUtil.empty
import org.tdf.sunflower.types.*
import java.io.File
import java.io.FileInputStream
import java.nio.file.Paths
import java.util.function.Function
import javax.servlet.http.HttpServletResponse

@RestController
@RequestMapping("/rpc")
class EntryController constructor(
    private val accountTrie: AccountTrie,
    private val peerServer: PeerServer,
    private val repo: RepositoryService,
) {

    private fun <T> getBlockOrHeader(
        rd: RepositoryReader, hashOrHeight: String, func: Function<Long, T>,
        func1: Function<HexBytes, T>
    ): T {
        var height: Long? = null
        try {
            height = hashOrHeight.toLong()
            val h = rd.bestHeader
            while (height < 0) {
                height += h.height + 1
            }
        } catch (ignored: Exception) {
        }
        if (height != null) {
            return func.apply(height)
        }
        return func1.apply(hashOrHeight.hex())
    }


    @GetMapping(value = ["/block/{hashOrHeight}"], produces = [MediaType.APPLICATION_JSON_VALUE])
    fun getBlock(@PathVariable hashOrHeight: String): BlockV1? {
        repo.reader.use { rd ->
            val b = getBlockOrHeader(rd, hashOrHeight, { rd.getCanonicalBlock(it) }, { rd.getBlockByHash(it) })
            return b?.v1()
        }
    }

    @GetMapping(value = ["/header/{hashOrHeight}"], produces = [MediaType.APPLICATION_JSON_VALUE])
    fun getHeader(@PathVariable hashOrHeight: String): HeaderV1? {
        repo.reader.use { rd ->
            val h = getBlockOrHeader(rd, hashOrHeight, { rd.getCanonicalHeader(it) }, { rd.getHeaderByHash(it) })
            return h?.v1()
        }
    }

    @GetMapping(value = ["/transaction/{hash}"], produces = [MediaType.APPLICATION_JSON_VALUE])
    fun getTransaction(@PathVariable hash: String): TransactionV1? {
        return repo.reader.use { it.getTransaction(hash.hex())?.v1() }
    }

    @GetMapping(value = ["/account/{addressOrPublicKey}"], produces = [MediaType.APPLICATION_JSON_VALUE])
    fun getAccount(@PathVariable addressOrPublicKey: String): AccountView {
        val addressHex = AddrUtil.of(addressOrPublicKey)
        repo.reader.use { rd ->
            val a = accountTrie.get(rd.bestHeader.stateRoot, addressHex) ?: Account.empty()
            return AccountView.fromAccount(addressHex, a)
        }
    }

    @GetMapping(value = ["/peers"], produces = [MediaType.APPLICATION_JSON_VALUE])
    fun peers(): PeersInfo {
        return PeersInfo(peerServer.peers, peerServer.bootstraps)
    }

    @GetMapping(value = ["/mstore8"], produces = [MediaType.TEXT_PLAIN_VALUE])
    fun getMSTORE8(resp: HttpServletResponse) {
        val f = File("MSTORE8")
        val has = f.exists()
        resp.writer.write(has.toString())
        resp.writer.close()
    }

    @GetMapping(value = ["/vm-logs/{id}"], produces = [MediaType.TEXT_PLAIN_VALUE])
    fun getLogs(@PathVariable id: String, resp: HttpServletResponse) {
        val p = AppConfig.get().vmLogs

        if (p.isEmpty()) {
            resp.writer.write("NO LOGS");
            resp.writer.close()
        }

        val path = Paths.get(p,  "$id.log")
        val f = FileInputStream(path.toString())
        IOUtils.copy(f, resp.outputStream)
    }

    data class PeersInfo(val peers: List<Peer>, val bootstraps: List<Peer>)

    data class AccountView(
        val address: HexBytes, // for normal account this field is continuous integer
        // for contract account this field is nonce of deploy transaction
        @JsonSerialize(using = IntSerializer::class) val nonce: Long, // the balance of account
        // for contract account, this field is zero
        @JsonSerialize(using = IntSerializer::class) val balance: Uint256, // for normal address this field is null
        // for contract address this field is creator of this contract
        val createdBy: HexBytes, // hash code of contract code
        // if the account contains none contract, contract hash will be null
        val contractHash: HexBytes, // root hash of contract db
        // if the account is not contract account, this field will be null
        val storageRoot: HexBytes
    ) {

        companion object {
            fun fromAccount(address: HexBytes, account: Account): AccountView {
                return AccountView(
                    address, account.nonce, account.balance,
                    empty(), account.contractHash,
                    account.storageRoot
                )
            }
        }
    }

    companion object {
        private val log = LoggerFactory.getLogger("rpc")
    }

    private fun Block.v1(): BlockV1 {
        return BlockV1.fromV2(this)
    }

    private fun Header.v1(): HeaderV1 {
        return HeaderV1.fromV2(this)
    }

    private fun Transaction.v1(): TransactionV1 {
        return TransactionV1.fromV2(this)
    }
}